### PR TITLE
Rule S4027: Exceptions should provide standard constructors

### LINKED ISSUE #384: Rule S4027: Exceptions should provide standard constructors
[RSPEC-4027](https://jira.sonarsource.com/browse/RSPEC-4027)

Exceptions types should provide the following constructors:

### PR DESCRIPTION
**Pull request by michalb-sonar**
Fix #384

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExceptionsNeedStandardConstructors.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                       KnownType.System_Runtime_Serialization_StreamingContext);
+        }
+
+        private static bool HasConstructor(ImmutableArray<IMethodSymbol> constructors,
+            Accessibility accessibility, params KnownType[] expectedParameterTypes)
+        {
+            return constructors.Any(c => AreParamTypesMatch(c, accessibility, expectedParameterTypes));
+        }
+
+        private static bool AreParamTypesMatch(IMethodSymbol constructor, Accessibility accessibility,
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExceptionsNeedStandardConstructors.cs]: This looks like an extension method already. I remember the same method in your previous PR...

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExceptionsNeedStandardConstructors.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                var classDeclaration = c.Node as ClassDeclarationSyntax;
+                var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+
+                if (classSymbol.DerivesFrom(KnownType.System_Exception) &&
+                    !HasStandardConstructors(classSymbol) &&
+                    !classDeclaration.Identifier.IsMissing)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExceptionsNeedStandardConstructors.cs]: I would move this one first as it seems to be the faster (no complex check).

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExceptionsNeedStandardConstructors.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            Accessibility accessibility, params KnownType[] expectedParameterTypes)
+        {
+            return constructors.Any(c => IsMatchingConstructor(c, accessibility, expectedParameterTypes));
+        }
+
+        private static bool IsMatchingConstructor(IMethodSymbol constructor, Accessibility accessibility,
+            KnownType[] expectedParameterTypes)
+        {
+            return constructor.DeclaredAccessibility == accessibility &&
+                   CollectionUtils.AreEqual(constructor.Parameters, expectedParameterTypes,
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExceptionsNeedStandardConstructors.cs]: Assuming this is an extension method I would use it as extension method :)

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Helpers/CollectionUtilsTest.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+ */
+
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using SonarAnalyzer.Helpers;
+using System.Collections.Generic;
+
+namespace SonarAnalyzer.UnitTest.Helpers
+{
+    [TestClass]
+    public class CollectionUtilsTest
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Helpers/CollectionUtilsTest.cs]: I would use FluentAssertions for all the tests here. Also use the extension method as an extension.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/PackagingTests/RuleTypeTests.cs [hunk 5]
```diff
@@ -389,7 +390,7 @@ private static void DetectTypeChanges(ResourceManager resourceManager, IImmutabl
             deletedRules.Should().BeEmpty($"YOU SHOULD NEVER DELETE RULES!");
 
             // IMPORTANT: If this test fails, you should update the types of the changed rules
-            // in the dictionaries above. Also add a GitHub issue specifying the change of type
+            // in the dictionaries above. Also add a GitHub issue specifying the change of type
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/PackagingTests/RuleTypeTests.cs]: Undo this change.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExceptionsNeedStandardConstructors.cs (hunk 3 above)
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExceptionsNeedStandardConstructors.cs]: It is not and I prefer it this way. I don't like how asymmetrical it is when doing a.Equals(b). Equals(a, b) reads much better.
